src/core/xml_parser.h

```diff
@@ -4,27 +4,107 @@
 #define XML_PARSER_MAX_ATTRIBUTES 13
 #define XML_PARSER_TAG_MAX_LENGTH 12
 
+/*
+ * @brief A structure that holds info about a xml element.
+ *
+ * Note: You can have many elements with the same name belonging to different parents.
+ */
 typedef struct {
-    const char *name;
-    int (*on_enter)(const char **attributes, int total_attributes);
-    void (*on_exit)(void);
-    const char *parent_names;
+    const char *name; /**< The name of the element. */
+    int (*on_enter)(void);     /**< A pointer to a function that should be called when the parser finds an element with this name. Can be NULL. */
+    void (*on_exit)(void);     /**< A pointer to a function that should be called when the parser gets to the end tag of the element with this name. Can be NULL. */
+    const char *parent_names;  /**< A list of valid parent names. If the element can have more than one parent, it should be separated by a "|"".*/
 } xml_parser_element;
 
+/**
+ * @brief Initiates the xml parser structure.
+ * 
+ * @param elements A list of xml_parser_element structures, indicating the format of the xml file. This will be used for error checking.
+ * @param total_elements The total number of elements passed.
+ * @return 1 if init was successful, 0 otherwise.
+ */
 int xml_parser_init(const xml_parser_element *elements, int total_elements);
 
+/**
+ * @brief Parses a chunk of a xml file. 
+ * 
+ * @param buffer The pointer to the xml string.
+ * @param buffer_size The size of the xml string.
+ * @param is_final Whether this is the last piece of the xml string. If you're passing a single string, this should be set to 1.
+ * @return 1 if parsing was successful, 0 otherwise.
+ */
 int xml_parser_parse(const char *buffer, int buffer_size, int is_final);
 
-int xml_parser_has_attribute(const char **attributes, const char *key);
-int xml_parser_get_attribute_int(const char **attributes, const char *key);
-const char *xml_parser_get_attribute_string(const char **attributes, const char *key);
-char *xml_parser_copy_attribute_string(const char **attributes, const char *key);
-int xml_parser_get_attribute_bool(const char **attributes, const char *key);
-int xml_parser_get_attribute_enum(const char **attributes, const char *key,
-    const char **values, int total_values, int start_offset);
+/**
+ * @brief Get the total number of attributes for the current element.
+ * 
+ * @return int The total number of attributes for the current element.
+ * Note that both keys and values are considered attributes, so if you just want to find the number of keys you should divide the result by 2.
```

Comment:
What if you have 2 keys with no values? Dividing by 2 would give 1 but the answer should be 2, right?

Reply:
Yes, but therec should be no Keys with no values as the xml standard doesn't support it.

If it happens though, there is no way for us to detect it,  but key1 will be the key and key2 the value.